design/design-doc-e2e-testing.md

```diff
@@ -0,0 +1,777 @@
+  In this way, users would be able to run these tests against their system
+  instances, either manually or in an automated way.
+
+* Parallel test execution.
+
+  End-to-end tests can take significantly more time to complete then unit tests,
+  and in order to keep CI jobs within acceptable limits, it should be possible
+  to execute tests in parallel against the same system instance
+
+* Run tests against an arbitrary system deployment.
+
+  In this way we can assure that Crossplane works properly in a number of k8s
+  clusters, deployed within a number of cloud providers
+
+* Test code reusability/maintainability/ease of development
+
+  Effort for adding a new test should decrease with the number of already
+  available tests, because we should be able to reuse the same/similar setup and
+  assertion logic. Similar to that, global changes in setup/assertion logic
+  should be kept in the single place ([DRY principle])
+
+As a part of this effort, already existing tests should be converted to the
+chosen approach/framework.
+
+## Non-Goals
+
+* Further followup actions should introduce more tests covering all Crossplane
+  core user-facing functionalities
+    * Package Manager
+        * Providers install/remove/update
+        * Configurations install/remove/update
+        * Error scenarios
+    * Composition Engine
+        * XRD createl/remove/update → verify effects on CRD
+        * Composition install/remove/update →effects on CompositionRevisions
+        * Create/update/delete a claim → verify the xr and mr and resource
+          readiness
+        * Patching → should have a patch of each type tested
+        * ConnectionDetails → verify the secret
+        * Late initialization → verify its propagated from mr->xr->claim
+        * Composition functions
+        * Composition validation
+        * ESS
+        * ObserveOnly scenarios
+        * Error scenarios
+    * Crossplane upgrades
+* Setting up CI to run e2e tests as a pre-merge checks
+
+These non-goals are important but out of the scope of this proposal.
+
+## Proposal
+
+We would like to propose the adoption of “Specification by Example”
+([wikipedia], [book]) methodology by defining tests using [Cucumber] framework
+and its [godog] bindings (2k stars on GitHub) for managing Crossplane
+end-to-end/acceptance tests. [Cucumber wiki] quote: _“Cucumber is a software
+tool that supports behavior-driven development (BDD). Central to the Cucumber
+BDD approach is its ordinary language parser called [Gherkin]. It allows
+expected software behaviors to be specified in a logical language that customers
+can understand. As such, Cucumber allows the execution of feature documentation
+written in business-facing text”. _
+
+Cucumber is a mature project/approach used successfully for testing systems of
+various complexity and size, for example:
+
+* [Service Binding Operator]
+* [Conformance tests for servicebinding specification]
+* [Primaza]
+* [Elastic end-to-end testing]
```

Comment:
Are we able to get a sense of whether choosing Cucumber resulted in tests that were significantly more useful to non-developers for these projects? How many tests for these projects were contributed by non-developers?

Reply:
It would be useful to add a section like this for the alternatives considered, too. i.e. What projects are using the alternatives.

Reply:
> Are we able to get a sense of whether choosing Cucumber resulted in tests that were significantly more useful to non-developers

I was directly involved in the first two listed project, and must say that choosing Cucumber brought all benefits listed in this proposal.